server/rest/src/main/java/org/infinispan/rest/resources/ServerResource.java

```diff
@@ -332,26 +344,82 @@ private CompletionStage<RestResponse> threads(RestRequest restRequest) {
-                                    Calendar.getInstance())
-                        )
-                        .entity(path.toFile()).build();
+                  return new NettyRestResponse.Builder().status(HttpResponseStatus.INTERNAL_SERVER_ERROR).build();
                }
+
+               return createReportResponse(path.toFile(), invocationHelper.getRestCacheManager().getNodeName());
             })
       );
    }
 
+   private RestResponse createReportResponse(Object report, String filename) {
+      return new NettyRestResponse.Builder()
+            .contentType(MediaType.fromString("application/gzip"))
+            .header("Content-Disposition",
+                  String.format("attachment; filename=\"%s-%s-%3$tY%3$tm%3$td%3$tH%3$tM%3$tS-report.tar.gz\"",
+                        Version.getBrandName().toLowerCase().replaceAll("\\s", "-"),
+                        filename,
+                        Calendar.getInstance())
+            )
+            .entity(report)
+            .build();
+   }
+
+   private CompletionStage<RestResponse> nodeReport(RestRequest restRequest) {
+      String targetName = restRequest.variables().get("nodeName");
+      EmbeddedCacheManager cacheManager = invocationHelper.getProtocolServer().getCacheManager();
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+      List<Address> members = cacheManager.getMembers();
+
+      if (invocationHelper.getRestCacheManager().getNodeName().equals(targetName)) {
+         return report(restRequest);
+      }
+
+      if (members == null) {
+         return CompletableFuture.completedFuture(responseBuilder.status(NOT_FOUND).build());
+      }
+
+      ByRef<ByteBuffer> response = new ByRef<>(null);
+      return SecurityActions.getClusterExecutor(cacheManager)
+            .submitConsumer(ecm -> {
+               CacheManagerInfo cmi = ecm.getCacheManagerInfo();
+               if (!cmi.getNodeName().equals(targetName)) return null;
+
+               GlobalComponentRegistry gcr = SecurityActions.getGlobalComponentRegistry(ecm);
+               BasicComponentRegistry bcr = gcr.getComponent(BasicComponentRegistry.class);
+               ServerStateManager ssm = bcr.getComponent(ServerStateManager.class).wired();
+
+               CompletionStage<byte[]> report = ssm.managedServer().getServerReport()
+                     .thenApply(path -> {
+                        try {
+                           return Files.readAllBytes(path);
+                        } catch (IOException e) {
+                           throw new CacheException(e);
+                        }
+                     });
+               return CompletableFutures.uncheckedAwait(report.toCompletableFuture());
+            }, (ignore, info, t) -> {
+               if (t != null) {
+                  throw CompletableFutures.asCompletionException(t);
+               }
+
+               if (info != null) {
+                  response.set(ByteBuffer.wrap(info));
```

Comment:
Why is there a need to wrap the array? Why not just store the array directly. I don't see anywhere we are using the ByteBuffer methods other than to just retrieve the array again.